Issue/2715 Extract Metadata from Data portal URL Solution using OpenFaas (Ready for Review)

t83714 opened the pull request:
### What this PR does

Fixes #2715 

This PR is based off PR: https://github.com/magda-io/magda/pull/2765 and make the functionality production-ready.

Change Summary & Deployment Info:
- More streamline & integration between Magda / openfaas deployment.
   - No extra step. No need to change current deployment command.
   - Openfaas core/gateway will be deployed to `[main Magda namespace]-openfaas` namespace (configurable)
     - e.g. if Magda is deployed to `default`, openfaas will be deployed to `default-openfaas` & `default-openfaas-fn`
   - Openfaas functions will be deployed to `[main Magda namespace]-openfaas-fn` namespace (configurable)
    - you can turn off openfaas namespaces auto-creation by set `magda-core.openfaas.createFunctionNamespace` or `magda-core.openfaas.createMainNamespace`. Please note, openfaas namespace requires proper labels set. Simply `kubectl create namespace` won't work
- **Use Helm 3 to deploy it**: The helm chart probably is still helm 2 deployable until PR for [this ticket](https://github.com/magda-io/magda/issues/2778) is landed, but you would have to take multiple steps (as it uses [helm 3's approach](https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#method-1-let-helm-do-it-for-you) to deploy it )
- Unlike helm 2, Helm 3 `--namespace` won't create the namespace if it does not exist. e.g. `helm upgrade --namespace magda` --- you need to create namespace `magda` before run it with helm 3. (openfaas namespaces will be auto created though)
- CRDs (CustomResourceDefinition): Helm3 will install it before all charts are installed. If it exists already, nothing will happen. 
- Openfaas gateway used to be exposed through gateway without auth. Now, by default, only admin users are allowed to access it (e.g. [this API](https://issue-2715-fin.dev.magda.io/api/v0/openfaas/system/functions)).  you can turn off auth by set `global.openfaas.allowAdminOnly`= false
- Scale to zero is turned on by default. the function pod will be scale to zero after 30 mins no usage. You can change this parameter via `magda-core.openfaas.faasIdler.inactivityDuration`="30m" .  **More available openfaas configuration option can be found [here](https://github.com/openfaas/faas-netes/tree/master/chart/openfaas#configuration)**
- helm 3 assume your command is namespaced. i.e. if you run `helm list` it will only list all releases in `default` namespace. if you want to delete the release, you need to specify the namespace (magda deployment namespace). e.g. `helm delete --namespace magda magda` openfaas namespaces will be auto-removed as well.
- ckan-connector: all JSON schema & js connector templates are asynchronously loaded at the same time (see [here](https://github.com/magda-io/magda-ckan-connector/blob/master/src/loadAspectBuilderFiles.ts)). This saves a few seconds of start-up time.  We still have some codebase (e.g. JsonConnector related code) are not `pure` and just import those libs (without running anything) might take up to 2 seconds --- I guess we can optimise it later (or migrate to node12 --- [seems some work has been done in this area](https://medium.com/@nodejs/introducing-node-js-12-76c41a1b3f3f) ). 

#### Test Site:

https://issue-2715-fin.dev.magda.io/

You need to login as admin in order to access the function.
Openfaas API is also protected by the auth.


### Checklist

-   [x] unit tests are included
-   [x] I've updated CHANGES.md with what I changed.
-   [x] I've linked this PR to an issue in ZenHub (core dev team only)

AlexGilleran commented on deploy/helm/openfaas/Chart.yaml:
```diff
@@ -0,0 +1,19 @@
+apiVersion: v1
+description: Enable Kubernetes as a backend for OpenFaaS (Functions as a Service)
```
Why did we need to copy-paste this in? Probably would be good to put the reason in here, or maybe in a readme.md file in this directory?

t83714 replied:
Mainly for:
- embedding it as part of Magda. e.g. control the namespace it's deployed into
- make it helm 3 compatible.
Will update the readme.md

t83714 replied:
Updated readme.md

AlexGilleran commented on magda-gateway/src/createOpenfaasGatewayProxy.ts:
```diff
@@ -0,0 +1,32 @@
+export default function createOpenfaasGatewayProxy(
+    options: OptionsType
+): express.Router {
+    const router = express.Router();
```
Could use a test for this - in particular to make sure that if anyone ever breaks this so that the `mustBeAdmin` thing stops working, it'll cause a build failure!

t83714 replied:
will add a test case.

t83714 replied:
Added test cases

AlexGilleran commented on magda-web-client/src/Components/Dataset/Add/DatasetLinkItem.tsx:
```diff
@@ -255,16 +260,117 @@ const DatasetLinkItemEditing = (props: EditViewProps) => {
     );
 };
 
+async function getAllDataUrlProcessorsFromOpenfaasGateway() {
```
Could use a JSDoc here - what's a data processor? How does one get it out of an open faas gateway?

t83714 replied:
Will add comments --- probably includes a url to openfaas gateway API swagger file

t83714 replied:
Added comments

AlexGilleran commented on magda-web-client/src/helpers/promiseAny.ts:
```diff
@@ -0,0 +1,41 @@
+                }
+                result[idx].isCompleted = true;
+                result[idx].error = e;
+                if (result.findIndex(item => !item.isCompleted) === -1) {
```
Could this be `!result.some(item => item.isCompleted)` ?

t83714 replied:
replace with `!result.some(item => !item.isCompleted)`

AlexGilleran commented on magda-web-client/src/helpers/promiseAny.ts:
```diff
@@ -0,0 +1,41 @@
+                if (result.findIndex(item => !item.isCompleted) === -1) {
+                    isResultReturned = true;
+                    reject(result.map(item => item.error).filter(item => item));
+                    return;
```
Not sure you need this return, it's the last line of the function?

t83714 replied:
removed

AlexGilleran commented on magda-web-client/src/helpers/promiseAny.ts:
```diff
@@ -0,0 +1,41 @@
+ * @param {Promise<any>[]} items
+ * @returns {Promise<any>}
+ */
+export default function promiseAny(items: Promise<any>[]): Promise<any> {
```
Not crazy about this name... how do you feel about something like `raceSuccessOnly`, seeing how it shares some of its semantics with `promise.race` ?

t83714 replied:
I actually try to replicate the functionality of this coming promise API (Promise.any):
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/any
but no preference on any particular names --- let me know if you prefer a `raceSuccessOnly` or we just leave as it is

AlexGilleran replied:
Ooh ok fair enough. Maybe mention that in the jsdoc then?

t83714 replied:
👍 That's a good point~ Have added more comments in the jsdoc

AlexGilleran replied:
:+1: I'll put that link the javadoc in my spelling changes commit :)

t83714 commented:
@AlexGilleran I think it's ready to have another look

t83714 commented:
@AlexGilleran 
Have added more comment to promiseAll.
Please have another look and let me know if anything else requires changes

AlexGilleran commented:
@t83714 OK so had a test - architecturally it's great! Testing through the API showed some problems though:

- https://data.gov.au/data/dataset/indigenous-land-use-agreements-registered-or-in-notification returned an error - "Not valid CKAN system URL" a number of times
- https://data.gov.au/data/dataset/0b1243c5-2cab-4675-b3f5-6b920daf377b doesn't work because it doesn't have any distributions (which makes sense), but it needs to tell the user this rather than just saying that it's invalid.

t83714 commented:
> * https://data.gov.au/data/dataset/indigenous-land-use-agreements-registered-or-in-notification

@AlexGilleran 
Fixed the bug and adjust the error message handling.
The demo site is also updated.
Please have a play when you have time
